### PR TITLE
`BigQuery`: support `encryptionConfiguration` in `google_bigquery_data_transfer_config`

### PR DESCRIPTION
**Pull request by BBBmau**
This PR would support the `encryptionConfiguration` field which is missing.

Related REST API Reference of `encryptionConfiguration`: https://cloud.google.com/bigquery/docs/reference/datatransfer/rest/v1/projects.locations.transferConfigs#encryptionconfiguration

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
bigquery: added `encryption_configuration` field to `google_bigquery_data_transfer_config` resource
```

### CODE REVIEW

File: mmv1/products/bigquerydatatransfer/Config.yaml [hunk 1]
```diff
@@ -172,6 +181,18 @@ properties:
       reingests data for [today-10, today-1], rather than ingesting data for
       just [today-1]. Only valid if the data source supports the feature.
       Set the value to 0 to use the default value.
+  - !ruby/object:Api::Type::NestedObject
+    name: 'encryptionConfiguration'
+    description: |
```
Comment (wj-chen) [inline mmv1/products/bigquerydatatransfer/Config.yaml:186]: To be future-proof, can we make this description more generic and not specific about encryption key?

Comment (BBBmau) [reply mmv1/products/bigquerydatatransfer/Config.yaml:186]: updated description to match that found in API Reference.

**Comment by BBBmau** [comment]
marking as ready for review with a reproducible tfconfig for manually testing this:

when applying the tfconfig below:

```hcl

data "google_project" "project" {
}

resource "google_kms_crypto_key" "mau_crypto_key" {
  name = "mau-crypto-key-test-bq"
  key_ring = "projects/hc-terraform-testing/locations/us-central1/keyRings/mau-latest-test-1"
  purpose = "ENCRYPT_DECRYPT"
}

data "google_iam_policy" "owner" {
  binding {
    role = "roles/bigquery.dataOwner"

    members = [
      "serviceAccount:mauricio-alvarezleon@hc-terraform-testing.iam.gserviceaccount.com",
    ]
  }
}

resource "google_bigquery_dataset_iam_policy" "dataset" {
  dataset_id  = google_bigquery_dataset.my_dataset.dataset_id
  policy_data = data.google_iam_policy.owner.policy_data
}

resource "google_bigquery_data_transfer_config" "query_config" {
  depends_on = [ google_kms_crypto_key.mau_crypto_key ]
  encryption_configuration {
    kms_key_name = "projects/hc-terraform-testing/locations/us-central1/keyRings/mau-latest-test/cryptoKeys/mau-crypto-key-test-bq"
  }
  display_name           = "my-query-mau"
  location               = "us-central1"
  data_source_id         = "scheduled_query"
  schedule               = "first sunday of quarter 00:00"
  destination_dataset_id = google_bigquery_dataset.my_dataset.dataset_id
  params = {
    destination_table_name_template = "my_table"
    write_disposition               = "WRITE_APPEND"
    query                           = "SELECT name FROM table WHERE x = 'y'"
  }
}

resource "google_bigquery_dataset" "my_dataset" {
  dataset_id    = "my_dataset"
  friendly_name = "foo"
  description   = "bar"
  location      = "us-central1"
}

```

We get a successfull apply with the following showing up for the `bigquery_transfer_config`, you'll notice that the encryption key has been successfully applied.

image

Let me know if anything else is needed to make this review easier. @wj-chen

File: mmv1/products/bigquerydatatransfer/Config.yaml (hunk 1 above)
Comment (wj-chen) [reply mmv1/products/bigquerydatatransfer/Config.yaml:186]: According to https://cloud.google.com/bigquery/docs/reference/datatransfer/rest/v1/projects.locations.transferConfigs#encryptionconfiguration - "The name of the KMS key used for encrypting BigQuery data." is the API description of `kmsKeyName`. Description of `EncryptionConfiguration` is "Represents the encryption configuration for a transfer."

**Comment by wj-chen** [comment]
> marking as ready for review with a reproducible tfconfig for manually testing this:
> 
> when applying the tfconfig below:
> 
> ```hcl
> data "google_project" "project" {
> }
> 
> resource "google_kms_crypto_key" "mau_crypto_key" {
>   name = "mau-crypto-key-test-bq"
>   key_ring = "projects/hc-terraform-testing/locations/us-central1/keyRings/mau-latest-test-1"
>   purpose = "ENCRYPT_DECRYPT"
> }
> 
> data "google_iam_policy" "owner" {
>   binding {
>     role = "roles/bigquery.dataOwner"
> 
>     members = [
>       "serviceAccount:mauricio-alvarezleon@hc-terraform-testing.iam.gserviceaccount.com",
>     ]
>   }
> }
> 
> resource "google_bigquery_dataset_iam_policy" "dataset" {
>   dataset_id  = google_bigquery_dataset.my_dataset.dataset_id
>   policy_data = data.google_iam_policy.owner.policy_data
> }
> 
> resource "google_bigquery_data_transfer_config" "query_config" {
>   depends_on = [ google_kms_crypto_key.mau_crypto_key ]
>   encryption_configuration {
>     kms_key_name = "projects/hc-terraform-testing/locations/us-central1/keyRings/mau-latest-test/cryptoKeys/mau-crypto-key-test-bq"
>   }
>   display_name           = "my-query-mau"
>   location               = "us-central1"
>   data_source_id         = "scheduled_query"
>   schedule               = "first sunday of quarter 00:00"
>   destination_dataset_id = google_bigquery_dataset.my_dataset.dataset_id
>   params = {
>     destination_table_name_template = "my_table"
>     write_disposition               = "WRITE_APPEND"
>     query                           = "SELECT name FROM table WHERE x = 'y'"
>   }
> }
> 
> resource "google_bigquery_dataset" "my_dataset" {
>   dataset_id    = "my_dataset"
>   friendly_name = "foo"
>   description   = "bar"
>   location      = "us-central1"
> }
> ```
> 
> We get a successfull apply with the following showing up for the `bigquery_transfer_config`, you'll notice that the encryption key has been successfully applied.
> 
> image
> 
> Let me know if anything else is needed to make this review easier. @wj-chen

Super helpful, thank you for the test notes!

File: mmv1/products/bigquerydatatransfer/Config.yaml (hunk 1 above)
Comment (BBBmau) [reply mmv1/products/bigquerydatatransfer/Config.yaml:186]: apologies seems like i got the descriptions mixed up, should be correct now.

**Review by wj-chen — APPROVED** [review]

File: mmv1/products/bigquerydatatransfer/Config.yaml [hunk 2]
```diff
@@ -46,13 +46,22 @@ examples:
     vars:
       display_name: 'my-query'
       dataset_id: 'my_dataset'
+  - !ruby/object:Provider::Terraform::Examples
+    name: 'bigquerydatatransfer_config_cmek'
+    skip_test: true
```
Comment (hao-nan-li) [inline mmv1/products/bigquerydatatransfer/Config.yaml:51]: Any particular reason to skip the test? If so it looks like the added field is not covered by any test.

**Review by hao-nan-li — COMMENTED** [review]
Thanks for adding this field. 
The Terraform google conversion error is unrelated to this change.

File: mmv1/products/bigquerydatatransfer/Config.yaml (hunk 2 above)
Comment (BBBmau) [reply mmv1/products/bigquerydatatransfer/Config.yaml:51]: This was based off of another example from `bigquery_dataset` which includes cmek support but doesn't include a test: https://github.com/GoogleCloudPlatform/magic-modules/pull/2226

I believe we skip it due to the test including a cryptoKeyRing creation, which we are unable to remove: https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings

I've included a test that includes the `kmsKeyName` field.

File: mmv1/third_party/terraform/services/bigquerydatatransfer/resource_bigquery_data_transfer_config_test.go [hunk 3]
```diff
@@ -383,6 +383,29 @@ func testAccBigqueryDataTransferConfig_scheduledQuery_update(t *testing.T) {
 	})
 }
 
+func testAccBigqueryDataTransferConfig_CMEK(t *testing.T) {
```
Comment (hao-nan-li) [inline mmv1/third_party/terraform/services/bigquerydatatransfer/resource_bigquery_data_transfer_config_test.go:386]: Any reason to skip the VCR run? Also I think this function should be named `TestAccBigqueryDataTransferConfig_CMEK`

Comment (BBBmau) [reply mmv1/third_party/terraform/services/bigquerydatatransfer/resource_bigquery_data_transfer_config_test.go:386]: same reason from above https://github.com/GoogleCloudPlatform/magic-modules/pull/11478#discussion_r1729437326

I also agree with the naming convention but went with what i committed due to the whole test file matching it.

Should we do the renaming of all the tests here or in a separate PR?

Comment (hao-nan-li) [reply mmv1/third_party/terraform/services/bigquerydatatransfer/resource_bigquery_data_transfer_config_test.go:386]: Ah thanks for letting me know. If other tests are created in this way then I'm fine with it.

**Comment by hao-nan-li** [comment]
Do you happen to know why this `contributor-membership-checker` is failing?

**Comment by BBBmau** [comment]
> Do you happen to know why this `contributor-membership-checker` is failing?

@hao-nan-li Seems like it's missing the build, i've reran the build to see if it'll get resolved by itself.

Relevant logs:
```hcl
request url:  https://api.github.com/repos/GoogleCloudPlatform/magic-modules/issues/11478
request body:  null

response status-code:  200

User is a core contributor
Error: Failed to find pending build for PR 11478
Usage:
  magician membership-checker [flags]
```

**Review by hao-nan-li — APPROVED** [review]

**Comment by SarahFrench** [comment]
Was the change to mmv1/.ruby-version intentional here?

**Pull request merged**